### PR TITLE
File-IO: IO Module and Open Builtin

### PR DESCRIPTION
**Pull request by rmliddle**
Start on some methods on the io Classes related to File-IO:  FileIO, BufferedReader/Writer and TextIOWrapper Classes and the open builtin.

Follows (approximately) the CPython implementation.

TextIOWrapper - BufferedReader/Writer - FileIO

Where open returns the TextIOWrapper or Buffered Type depending on the mode parameters. 

Some related implementation from the os module, memoryview and minor changes to bytes and byte array types.

[open](https://docs.python.org/3/library/functions.html#open) builtin at this stage only supports first two parameters: file and mode

### CODE REVIEW

**Comment by rmliddle** [comment]
Related to #220 and #243

File: vm/src/stdlib/io.rs [hunk 1]
```diff
@@ -29,17 +59,370 @@ fn bytes_io_getvalue(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
...
+
+    if modes.len() == 0 || modes.len() > 1 {
+        return Err(vm.new_value_error("Invalid Mode".to_string()));
+    }
+
+    //Class objects (potentially) consumed by io.open
+    //RawIO: FileIO
+    //Buffered: BufferedWriter, BufferedReader
+    //Text: TextIOWrapper
+    let file_io_class = vm.ctx.get_attr(&module, "FileIO").unwrap();
```
Comment (windelbouwman) [inline vm/src/stdlib/io.rs]: Could you use here instead of `unwrap();` the question mark operator `?;`?

**Comment by windelbouwman** [comment]
Fantastic! This is a major milestone! I have some minor comments, it looks very good though!

File: vm/src/pyobject.rs [hunk 2]
```diff
@@ -732,6 +748,21 @@ impl DictProtocol for PyObjectRef {
...
 }
 
+pub trait BufferProtocol {
+    fn readonly(&self) -> bool;
+}
+
+impl BufferProtocol for PyObjectRef {
+    fn readonly(&self) -> bool {
+        match self.borrow().kind {
+            PyObjectKind::Bytes { value: _ } => false,
```
[the remark this answers is not in the document]
Comment (rmliddle) [reply vm/src/pyobject.rs]: This certainly makes a lot of sense in retrospect!

I've made the change here to match on the type name string instead.

**Pull request merged**